File: sdk/flutter/lib/src/rewarded_ad.dart
// ---------------------------------------------------------------------------
// LevelMomentRewardedAd
//
// Drop-in replacement for AdMob's RewardedAd. The SDK is a thin shell over a
// WebView pointing at the hosted break page (platform/web/app/break). The page
// renders all question UI, fetches data, submits answers, AND owns the
// impression queue. Games call show(context:) and handle onUserEarnedReward /
// onAdDismissedFullScreenContent only. See docs/ADR-001-webview-rendering.md.
//
// Structurally mirrors sdk/react-native/src/LevelMomentAd.ts.
//
// BEFORE (AdMob):
//   RewardedAd.load(
//     adUnitId: 'ca-app-pub-xxx/yyy',
//     request: const AdRequest(),
//     rewardedAdLoadCallback: RewardedAdLoadCallback(
//       onAdLoaded: (ad) => _rewardedAd = ad,
//       onAdFailedToLoad: (err) => retry(),
//     ),
//   );
//   _rewardedAd?.fullScreenContentCallback = FullScreenContentCallback(
//     onAdDismissedFullScreenContent: (ad) { ad.dispose(); resumeGame(); },
//   );
//   _rewardedAd?.show(onUserEarnedReward: (ad, reward) => grantBonus());
//
// AFTER (LevelMoment — identical structure, the hosted page renders everything):
//   LevelMomentRewardedAd.load(
//     placementId: 'your-game-id',
//     adLoadCallback: LevelMomentAdLoadCallback(
//       onAdLoaded: (ad) => _rewardedAd = ad,
//       onAdFailedToLoad: (err) => retry(),
//     ),
//   );
//   _rewardedAd?.fullScreenContentCallback = LevelMomentFullScreenContentCallback(
//     onAdDismissedFullScreenContent: (ad) { ad.dispose(); resumeGame(); },
//   );
//   _rewardedAd?.show(
//     context: context,
//     onUserEarnedReward: (ad, reward) => grantBonus(),
//   );
// ---------------------------------------------------------------------------

import 'package:flutter/material.dart';

import 'constants.dart';
import 'credential_bridge.dart';
import 'levelmoment_ads.dart';
import 'models.dart';
import 'widgets/level_moment_web_view.dart';

class LevelMomentRewardedAd {
  /// Mirrors: adUnitId — your game's identifier from the LevelMoment developer portal
  final String placementId;

  /// A credential to start from. Normally null: a paired device's credential
  /// is held by the SDK (platform secure store) and by the hosted page, and
  /// neither needs the game to supply it. Set it only for a sandbox token
  /// while you are integrating, or a token you read yourself from a
  /// parent-portal link. It is handed to the page over the bridge, never on
  /// the URL.
  final String? studentToken;

  /// SSV-parity custom data (mirrors LevelMomentConfig.customData in sdk-core).
  /// Sent in the credential handshake so the page stamps it on every
  /// impression it records and echoes it on reward.earned.
  final String? customData;

  /// Mirrors: fullScreenContentCallback — set this before calling show()
  LevelMomentFullScreenContentCallback? fullScreenContentCallback;

  // Internal state
  final String _format;
  bool _loaded = false;
  bool _disposed = false;

  LevelMomentRewardedAd._({
    required this.placementId,
    required String format,
    this.studentToken,
    this.customData,
  }) : _format = format;

  // ---------------------------------------------------------------------------
  // Static factory — mirrors RewardedAd.load()
  // ---------------------------------------------------------------------------

  /// Mark the ad ready to show. The actual question fetch happens inside the
  /// WebView when show() is called — there is no separate native preload step
  /// in this architecture. Kept as a separate call so consumers can stay on
  /// the familiar AdMob load -> show pattern.
  ///
  /// Mirrors: RewardedAd.load(adUnitId, request, rewardedAdLoadCallback) and
  /// LevelMomentAd.load() in the react-native SDK (synchronous mark-ready, no
  /// network).
  static Future<void> load({
    required String placementId,
    required LevelMomentAdLoadCallback adLoadCallback,
    String? studentToken,
    String format = 'flashcard',
    String? customData,
  }) async {
    assert(
      LevelMomentAds.instance.isInitialized,
      'Call LevelMomentAds.instance.initialize() before loading ads.',
    );

    String? resolvedToken;
    try {
      resolvedToken = LevelMomentAds.instance.resolveStudentToken(studentToken);
    } catch (err) {
      adLoadCallback.onAdFailedToLoad(
        LevelMomentAdError(code: 'invalid_request', message: '$err'),
      );
      return;
    }
    final ad = LevelMomentRewardedAd._(
      placementId: placementId,
      studentToken: resolvedToken,
      format: format,
      customData: customData,
    );
    ad._loaded = true;
    adLoadCallback.onAdLoaded(ad);
  }

  // ---------------------------------------------------------------------------
  // Instance API — mirrors the RewardedAd object returned to onAdLoaded
  // ---------------------------------------------------------------------------

  bool get isLoaded => _loaded && !_disposed;

  /// The break format: 'flashcard', 'quiz', or 'deep_dive'.
  String get format => _format;

  /// Display the break. No network call here — the hosted page does the fetch.
  /// Pushes a fullscreen route hosting a WebView that points at the hosted
  /// /break page, which renders all question types and session flows, then
  /// posts back terminal events.
  ///
  /// Mirrors: rewardedAd.show(onUserEarnedReward: ...) and LevelMomentAd.show().
  void show({
    required BuildContext context,
    required void Function(
            LevelMomentRewardedAd ad, LevelMomentRewardItem reward)
        onUserEarnedReward,
  }) {
    if (!isLoaded) {
      fullScreenContentCallback?.onAdFailedToShowFullScreenContent?.call(
        this,
        const LevelMomentAdError(
          code: 'not_loaded',
          message: 'show() called before onAdLoaded. Call load() first.',
        ),
      );
      return;
    }

    // Terminal-once guard for the dismiss path: earnedReward may fire many
    // times; dismissed/error fire the dismissal exactly once. Mirrors the
    // _shown/dismissedRef discipline in the react-native SDK.
    var terminal = false;

    void handleMessage(HostMessage message) {
      // Keep the secure store in step with the page: store what pairing
      // minted, forget what the server refused. The WebView answers
      // `needCredential` itself — it holds the controller that can inject.
      if (applyCredentialMessage(
        message,
        placementId,
        useDeviceStore: !LevelMomentAds.instance.mock &&
            LevelMomentAds.instance.unsafeTesting == null,
      )) {
        return;
      }
      switch (message) {
        // `openExternal` is launched by the WebView itself; the break carries
        // on behind the browser.
        case NeedCredential() ||
              CredentialIssued() ||
              CredentialInvalid() ||
              OpenExternal():
          return;
        case Ready():
          fullScreenContentCallback?.onAdShowedFullScreenContent?.call(this);
        case EarnedReward(:final amount, :final rewardId):
          onUserEarnedReward(
            this,
            LevelMomentRewardItem(
              type: 'question_answered',
              amount: amount,
              rewardId: rewardId,
            ),
          );
        // SignedIn belongs to the sign-in gate and never reaches a break. If
        // one ever arrives the surface has closed, so resume the game rather
        // than leaving it waiting for a dismiss that will not come.
        case Dismissed() || SignedIn():
          if (terminal) return;
          terminal = true;
          fullScreenContentCallback?.onAdDismissedFullScreenContent?.call(this);
          dispose();
        case ErrorMsg(:final code, :final message):
          if (terminal) return;
          terminal = true;
          fullScreenContentCallback?.onAdFailedToShowFullScreenContent?.call(
            this,
            LevelMomentAdError(code: code, message: message),
          );
          dispose();
      }
    }

    final hostedUrl = buildUrl();
    Navigator.of(context).push(
      MaterialPageRoute<void>(
        fullscreenDialog: true,
        builder: (_) => LevelMomentWebView(
          key: ValueKey(hostedUrl),
          url: hostedUrl,
          onMessage: handleMessage,
          onNeedCredential: credentialResponder(
            placementId: placementId,
            explicitToken: studentToken,
            customData: customData,
            useDeviceStore: !LevelMomentAds.instance.mock &&
                LevelMomentAds.instance.unsafeTesting == null,
            origin: levelMomentOrigin(LevelMomentAds.instance.breakUrl),
          ),
        ),
      ),
    );
  }

  /// Release resources. Called automatically after the break ends.
  /// Mirrors: ad.dispose()
  void dispose() {
    _disposed = true;
  }

  // ---------------------------------------------------------------------------
  // URL building — mirrors LevelMomentAd._buildUrl() in the react-native SDK
  // ---------------------------------------------------------------------------

  /// Build the hosted /break page URL with placement / format params.
  /// Exposed for testing; treat as private elsewhere.
  @visibleForTesting
  String buildUrl() {
    final params = <String, String>{
      'placementId': placementId,
      'format': _format,
    };
    if (LevelMomentAds.instance.mock) {
      params['mock'] = 'true';
    } else {
      params['apiUrl'] = LevelMomentAds.instance.apiUrl;
    }
    // No credential rides on this URL. The page asks over the bridge
    // (`needCredential`) and the answer comes from the secure store, so a live
    // token never reaches a launch URL, a crash report, or a web log.
    params[kHostCapabilitiesParam] = kHostCapabilities;
    params['protocolVersion'] = '$kLevelMomentProtocolVersion';
    params['sdkVersion'] = kLevelMomentSdkVersion;
    if (LevelMomentAds.instance.isUnsafeTesting) {
      params['sandbox'] = 'true';
    }

    final query = params.entries
        .map((e) =>
            '${Uri.encodeQueryComponent(e.key)}=${Uri.encodeQueryComponent(e.value)}')
        .join('&');
    final breakUrl = LevelMomentAds.instance.breakUrl;
    final sep = breakUrl.contains('?') ? '&' : '?';
    return '$breakUrl$sep$query';
  }
}

File: sdk/unity/Runtime/RewardedAd.cs
// ---------------------------------------------------------------------------
// LevelMoment Unity SDK — RewardedAd (ADR-001 WebView shell).
//
// Drop-in shape for AdMob's RewardedAd: Load() prepares, Show() presents. The
// SDK renders nothing — Show() opens the hosted /break page in a WebView and
// bridges the page's postMessage events (ready / earnedReward / dismissed /
// error) to the C# callbacks. Structurally mirrors sdk/flutter's
// LevelMomentRewardedAd and sdk/react-native's LevelMomentAd:
//   - Load() is a synchronous mark-ready — no network (the page does the fetch).
//   - dismissed/error are terminal and collapse to a single dismiss (the
//     terminal-once guard); earnedReward may fire many times before it.
//   - A pre-`ready` load watchdog (15s, mirroring sdk/web) prevents a crashed
//     page from covering the game forever.
//
// The WebView lifecycle, watchdog, and terminal-once collapse live in
// BreakSurfaceCore, shared with InterstitialAd (see InterstitialAd.cs). This
// file is left owning only what is specific to a rewarded break: the reward
// dispatch, and the public Load/Show/Destroy shape a game developer sees.
//
// See docs/ADR-001-webview-rendering.md.
// ---------------------------------------------------------------------------

using System;

namespace LevelMoment
{
    public class RewardedAd
    {
        /// <summary>
        /// EditMode tests set this true to skip creating the runtime
        /// MonoBehaviour driver; they drive Tick() manually with an injected
        /// clock instead. Shared with InterstitialAd and the sign-in gate.
        /// </summary>
        internal static bool SkipRuntimeDriver
        {
            get { return LevelMomentRuntime.SkipDriver; }
            set { LevelMomentRuntime.SkipDriver = value; }
        }

        // No `kind` is ever passed to the core here — the /break URL this
        // placement opens is byte-for-byte what it always was. See
        // InterstitialAd, which passes "interstitial".
        private readonly BreakSurfaceCore<RewardedAdShowCallbacks> _core;

        private RewardedAd(string placementId, string format, string studentToken)
        {
            _core = new BreakSurfaceCore<RewardedAdShowCallbacks>(
                placementId, format, null, studentToken, HandleEarnedReward);
        }

        // ---- Static factory — mirrors RewardedAd.Load() ---------------------

        /// <summary>
        /// Mark an ad ready to Show(). Synchronous, no network — the hosted
        /// page fetches the question when Show() opens it. No token: a paired
        /// device's credential lives on the Level Moment origin, and the
        /// hosted page finds it there. Use the overload only for a sandbox
        /// token while you integrate.
        /// </summary>
        public static void Load(
            string placementId,
            RewardedAdLoadCallbacks callbacks,
            string format = "quick_question")
        {
            Load(placementId, null, callbacks, format);
        }

        /// <summary>
        /// Load with an explicit student session token — a sandbox token, for
        /// integration testing. A paired device needs none. Nothing reads a
        /// credential off the launch URL: a URL lands in history, in the referrer
        /// of whatever loads next, and in any crash report taken afterwards.
        /// </summary>
        public static void Load(
            string placementId,
            string studentToken,
            RewardedAdLoadCallbacks callbacks,
            string format = "quick_question")
        {
            string resolvedToken;
            LevelMomentAdError error;
            if (!BreakSurfaceCore.TryResolveLoad(
                placementId, studentToken,
                "Call LevelMomentAds.Initialize() before RewardedAd.Load().",
                out resolvedToken, out error))
            {
                Fail(callbacks, error);
                return;
            }

            var ad = new RewardedAd(placementId, format, resolvedToken);
            ad._core.MarkLoaded();

            if (callbacks != null && callbacks.OnAdLoaded != null)
                callbacks.OnAdLoaded(ad);
        }

        private static void Fail(RewardedAdLoadCallbacks callbacks, LevelMomentAdError error)
        {
            if (callbacks != null && callbacks.OnAdFailedToLoad != null)
                callbacks.OnAdFailedToLoad(error);
        }

        // ---- Instance API ---------------------------------------------------

        /// <summary>True once loaded and not yet shown/disposed.</summary>
        public bool IsLoaded
        {
            get { return _core.IsLoaded; }
        }

        /// <summary>The break format: <c>quick_question</c>, <c>practice_set</c>, <c>mastery_round</c>, or <c>intro_lesson</c>.</summary>
        public string Format
        {
            get { return _core.Format; }
        }

        /// <summary>
        /// Present the break. No network here — the hosted page does the fetch.
        /// Opens a fullscreen WebView at the hosted /break page and bridges its
        /// events to <paramref name="callbacks"/>.
        ///
        /// Catches WebView-provider startup failures. If the provider factory or
        /// <c>Open()</c> throws, the SDK reports <c>webview_error</c> through
        /// <c>OnAdFailedToShow</c> and tears the ad down, so the game resumes on
        /// its own callback. Exceptions thrown by your own callbacks propagate
        /// to you unchanged.
        /// </summary>
        public void Show(RewardedAdShowCallbacks callbacks)
        {
            _core.Show(callbacks);
        }

        /// <summary>
        /// Release the ad. Tears down the WebView without firing callbacks (the
        /// game is already handling teardown). Mirrors AdMob's Destroy().
        /// </summary>
        public void Destroy()
        {
            _core.Destroy();
        }

        // ---- Watchdog tick (driven by LevelMomentRuntime, or tests) ---------

        internal void Tick()
        {
            _core.Tick();
        }

        // ---- Reward dispatch (the one piece of bridge handling that is
        // specific to a rewarded break) --------------------------------------

        private void HandleEarnedReward(HostMessage msg)
        {
            var callbacks = _core.Callbacks;
            if (callbacks == null)
                return;

            if (callbacks.OnUserEarnedRewardItem != null)
            {
                var earnedItem = callbacks.OnUserEarnedRewardItem;
                var item = new LevelMomentRewardItem { Amount = msg.Amount, RewardId = msg.RewardId };
                _core.FireToPublisher(delegate { earnedItem(item); });
            }
            if (!_core.IsTerminal && callbacks.OnUserEarnedReward != null)
            {
                var earned = callbacks.OnUserEarnedReward;
                var amount = msg.Amount;
                _core.FireToPublisher(delegate { earned(amount); });
            }
        }
    }
}
